File: src/main/java/org/distril/beengine/material/item/ItemPalette.java
package org.distril.beengine.material.item;

import com.google.common.collect.BiMap;
import com.google.common.collect.HashBiMap;
import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.nukkitx.protocol.bedrock.data.inventory.ItemData;
import com.nukkitx.protocol.bedrock.packet.CreativeContentPacket;
import com.nukkitx.protocol.bedrock.packet.StartGamePacket;
import org.distril.beengine.Bootstrap;
import org.distril.beengine.util.ItemUtils;

import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class ItemPalette {

	private static final Gson GSON = new Gson();

	private static final BiMap<String, Integer> ITEMS = HashBiMap.create();
	private static final Set<StartGamePacket.ItemEntry> ENTRIES = new HashSet<>();

	private static final CreativeContentPacket CREATIVE_CONTENT_PACKET = new CreativeContentPacket();

	static {
		try (var reader = new InputStreamReader(Bootstrap.getResource("data/runtime_item_states.json"))) {
			var items = GSON.fromJson(reader, JsonArray.class);

			items.forEach(item -> {
				var itemObj = item.getAsJsonObject();
				var identifier = itemObj.get("name").getAsString();
				var runtimeId = itemObj.get("id").getAsInt();
				ITEMS.put(identifier, runtimeId);
				ENTRIES.add(new StartGamePacket.ItemEntry(identifier, (short) runtimeId, false));
			});
		} catch (IOException exception) {
			throw new RuntimeException(exception);
		}

		try (var reader = new InputStreamReader(Bootstrap.getResource("data/creative_items.json"))) {
			var itemArray = GSON.fromJson(reader, JsonObject.class).getAsJsonArray("items");

			List<ItemData> itemsData = new ArrayList<>();
			int netId = 0;
			for (var itemElement : itemArray) {
				var itemObj = itemElement.getAsJsonObject();

				netId++;
				itemsData.add(
						ItemUtils.fromJSON(itemObj)
								.toBuilder()
								.netId(netId)
								.build()
				);
			}


			CREATIVE_CONTENT_PACKET.setContents(itemsData.toArray(new ItemData[0]));
		} catch (IOException exception) {
			throw new RuntimeException(exception);
		}
	}

	public static List<StartGamePacket.ItemEntry> getItemEntries() {
		return ENTRIES.stream().toList();
	}

	public static CreativeContentPacket getCreativeContentPacket() {
		return CREATIVE_CONTENT_PACKET;
	}

	public static String getIdentifier(int runtimeId) {
		return ITEMS.inverse().get(runtimeId);
	}

	public static int getRuntimeId(String identifier) {
		return ITEMS.get(identifier);
	}
}

File: src/main/kotlin/org/distril/beengine/material/item/ItemPalette.kt
package org.distril.beengine.material.item

import com.google.gson.JsonArray
import com.google.gson.JsonObject
import com.nukkitx.protocol.bedrock.data.inventory.ItemData
import com.nukkitx.protocol.bedrock.packet.CreativeContentPacket
import com.nukkitx.protocol.bedrock.packet.StartGamePacket
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import org.distril.beengine.util.ItemUtils
import org.distril.beengine.util.Utils
import org.distril.beengine.util.Utils.gson
import java.io.InputStreamReader

object ItemPalette {

	val items = mutableMapOf<String, Int>()
	val inverseItems: Map<Int, String>
		get() = items.entries.associate { (k, v) -> v to k }

	val entries = mutableSetOf<StartGamePacket.ItemEntry>()

	val creativeItems = mutableMapOf<Int, ItemData>()

	val creativeContentPacket = CreativeContentPacket()

	suspend fun init() = coroutineScope {
		launch {
			InputStreamReader(Utils.getResource("data/runtime_item_states.json")).use {
				val itemsJsons = gson.fromJson(it, JsonArray::class.java)
				itemsJsons.forEach { element ->
					val itemJson = element.asJsonObject
					val identifier = itemJson["name"].asString
					val runtimeId = itemJson["id"].asInt

					items[identifier] = runtimeId

					entries.add(StartGamePacket.ItemEntry(identifier, runtimeId.toShort(), false))
				}
			}
		}

		launch {
			InputStreamReader(Utils.getResource("data/creative_items.json")).use {
				val itemsJsons = gson.fromJson(it, JsonObject::class.java).getAsJsonArray("items")
				var networkId = 0
				itemsJsons.forEach { element ->
					networkId++

					val itemJson = element.asJsonObject
					creativeItems[networkId] = ItemUtils.fromJSON(itemJson).toBuilder().netId(networkId).build()
				}

				creativeContentPacket.contents = creativeItems.values.toTypedArray()
			}
		}
	}

	fun getIdentifier(runtimeId: Int) = inverseItems[runtimeId]!!

	fun getRuntimeId(identifier: String) = items[identifier]!!

	fun getCreativeItem(itemNetworkId: Int) = ItemUtils.fromNetwork(creativeItems[itemNetworkId])
}
